qiskit/transpiler/passes/synthesis/unitary_synthesis.py

```diff
@@ -82,31 +103,90 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
         Returns:
             Output dag with UnitaryGates synthesized to target basis.
         """
+        
         euler_basis = _choose_euler_basis(self._basis_gates)
         kak_gate = _choose_kak_gate(self._basis_gates)
-
         decomposer1q, decomposer2q = None, None
         if euler_basis is not None:
             decomposer1q = one_qubit_decompose.OneQubitEulerDecomposer(euler_basis)
         if kak_gate is not None:
-            decomposer2q = TwoQubitBasisDecomposer(kak_gate, euler_basis=euler_basis)
-
-        for node in dag.named_nodes("unitary"):
+            decomposer2q = TwoQubitBasisDecomposer(
+                kak_gate, euler_basis=euler_basis, pulse_optimize=self._pulse_optimize
+            )
 
+        for node in dag.named_nodes("unitary", "swap"):
             synth_dag = None
+            wires = None
             if len(node.qargs) == 1:
                 if decomposer1q is None:
                     continue
                 synth_dag = circuit_to_dag(decomposer1q._decompose(node.op.to_matrix()))
             elif len(node.qargs) == 2:
                 if decomposer2q is None:
                     continue
-                synth_dag = circuit_to_dag(
-                    decomposer2q(node.op.to_matrix(), basis_fidelity=self._approximation_degree)
-                )
+                synth_dag, wires = self._synth_natural_direction(node, wires, dag, decomposer2q)
             else:
                 synth_dag = circuit_to_dag(isometry.Isometry(node.op.to_matrix(), 0, 0).definition)
 
-            dag.substitute_node_with_dag(node, synth_dag)
+            dag.substitute_node_with_dag(node, synth_dag, wires=wires)
 
         return dag
+
+    def _synth_natural_direction(self, node, wires, dag, decomposer2q):
+        layout = self.property_set["layout"]
+        natural_direction = None
+        synth_direction = None
+        physical_gate_fidelity = None
+        if self._natural_direction and layout and self._coupling_map:
+            neighbors = self._coupling_map.neighbors(dag.qubits.index(node.qargs[0]))
+            zero_one = dag.qubits.index(node.qargs[1]) in neighbors
+            neighbors = self._coupling_map.neighbors(dag.qubits.index(node.qargs[1]))
+            one_zero = dag.qubits.index(node.qargs[0]) in neighbors
+            if zero_one and not one_zero:
+                natural_direction = [0, 1]
+            if one_zero and not zero_one:
+                natural_direction = [1, 0]
+
+        elif self._natural_direction and layout and self._backend_props:
+            len_0_1 = inf
+            len_1_0 = inf
+            try:
+                len_0_1 = self._backend_props.gate_length(
+                    "cx", [dag.qubits.index(node.qargs[0]), dag.qubits.index(node.qargs[1])]
```

Comment:
Rather than hard-coding `"cx"`, you might look up the name of the `.gate` field on `decomposer2q`.